earth_moving/clutter/engine2.py

```diff
@@ -0,0 +1,346 @@
+# this file contains the PyBulletEnvironment class
+# the general capabilities are opening the environment, closing the environment, and general control functions
+
+# general imports
+import pybullet as p
+import pybullet_data
+import time as t
+import numpy as np
+import math
+
+# class definition
+class PyBulletEnvironment:
+    
+    # constructor
+    def __init__(self, gui=True, gravity=(0, 0, -10)):
+        """
+        Initialize the PyBullet environment.
+
+        :param gui: Boolean, if True the environment will be opened with a GUI.
+        :param gravity: Tuple, the gravity vector.
+        """
+        
+        # gui
+        self.gui = gui
+        self.physicsClient = None 
+        
+        # control accuracy threshold        
+        self.control_threshold = 1e-2 
+        
+        # iter limit
+        self.iter_limit = 5e2
+        
+        # ID list 
+        self.ID = []
+        
+        # pybullet options
+        self.URDF_MERGE_FIXED_LINKS = True
+        self.URDF_USE_INERTIA_FROM_FILE = True
+        self.URDF_USE_SELF_COLLISION = True
+        self.URDF_USE_SELF_COLLISION_INCLUDE_PARENT = True
+        self.URDF_USE_SELF_COLLISION_EXCLUDE_ALL_PARENTS = False
+        self.URDF_USE_IMPLICIT_CYLINDER = True
+        self.URDF_ENABLE_SLEEPING = False
+        self.URDF_INITIALIZE_SAT_FEATURES = False
+        self.URDF_USE_MATERIAL_COLORS_FROM_MTL = False
+        self.URDF_ENABLE_CACHED_GRAPHICS_SHAPES = False
+        self.URDF_MAINTAIN_LINK_ORDER = True
+        
+        # put in a bitwise OR the previous flags to combine them
+        self.flags =    self.URDF_MERGE_FIXED_LINKS | self.URDF_USE_INERTIA_FROM_FILE | self.URDF_USE_SELF_COLLISION | \
+                        self.URDF_USE_SELF_COLLISION_INCLUDE_PARENT | self.URDF_USE_SELF_COLLISION_EXCLUDE_ALL_PARENTS | \
+                        self.URDF_USE_IMPLICIT_CYLINDER | self.URDF_ENABLE_SLEEPING | self.URDF_INITIALIZE_SAT_FEATURES | \
+                        self.URDF_USE_MATERIAL_COLORS_FROM_MTL | self.URDF_ENABLE_CACHED_GRAPHICS_SHAPES | self.URDF_MAINTAIN_LINK_ORDER
+        
+        
+        # set the gravity
+        self.gravity = gravity
+
+    # open environment
+    def open_environment(self, robot_urdf="robot.urdf", init_pos=(0, 0, 0), init_quat=(0, 0, 0, 1)):
```

Comment:
Need to think: what does "environment" represent? perhaps it's a different class? Think what happens when we want to launch the same setup but with mujoco. Let's talk about that...